### ISSUE TITLE
Make private fun Instant.atZone(zone: TimeZone): ZonedDateTime public

### ISSUE THREAD

**Issue by andylamax** [Kotlin/kotlinx-datetime#82]
Change from
```kotlin
private fun Instant.atZone(zone: TimeZone): ZonedDateTime = value.atZone(zone.zoneId)
```

To
```kotlin
public fun Instant.atZone(zone: TimeZone): ZonedDateTime = value.atZone(zone.zoneId)
```

**Comment by dkhalanskyjb** [comment]
To do this, we would first need to make `ZonedDateTime` available on all platforms. However, we didn't see enough compelling use cases for `ZonedDateTime` to decide which API it should have and whether we want to add this class at all. Could you please share your use cases for `Instant.atZone`?

**Comment by andylamax** [comment]
I need to be able to get an `Instant` at a particular `TimeZone`. Calling `Clock.System.now()` returns an `Instant` at `TimeZone.UTC`, If I need to display the instant at another timezone say `TimeZone.GMT(3)`, I will have to do

```kotlin
val now = Clock.System.now()
val nowLocal = now.toLocalDateTime(TimeZone.GMT(3)).toInstant(TimeZone.GMT(3))
```

Which is kind of tedious. don't you think?

**Comment by dkhalanskyjb** [comment]
Getting an `Instant` at a particular `TimeZone` doesn't really make sense because `Instant` represents location-independent moments in time. It can be used to answer questions like "How many hours is it until that concert starts?" It doesn't matter in which time zone the question is asked, the answer is the same.

For example, the code you provided is functionally exactly the same as
```kotlin
val now = Clock.System.now()
val nowLocal = now
```

The reason is, `now.toLocalDateTime(TimeZone.GMT(3))` essentially asks "What is the local representation of this absolute moment in time?", and `localDateTime.toInstant(TimeZone.GMT(3))` asks "What absolute moment in time corresponds to this local representation?" The two operations cancel each other out for fixed-offset time zones. For time zones with DST, `nowLocal` would be equal to `now` in literally 99.99% of cases and would have a strange value in the remaining 0.01%.

You do raise a good point about needing to *display* an instant in a local time zone! We do know of this use case, but it seems odd to implement a whole `ZonedDateTime` class for formatting only. Instead we could, just as an example, simply add an optional `TimeZone` parameter to `Instant.toString()`.

**Comment by andylamax** [comment]
Thank you for your graciously explanation.

Just for clarification, please answer me this. What do you mean you say 'Instant' represents location independent moments in time. Does that roughly translate to a representation of moment at `GMT=0`?

Also, based on your explanation, I totally understand when `now.toLocalDateTime(TimeZone.GMT(3))` asks "What is the local representation of this absolute moment in time?". However I am having a hard time grasping why we still need to pass a `TimeZone` when asking the question "What is the *absolute moment* moment in time". If it is absolute, it is independent of the `TimeZone`. Am I contradicting my self? Not sure

**Comment by dkhalanskyjb** [comment]
> Does that roughly translate to a representation of moment at GMT=0?

Almost, though not quite. An `Instant` is implemented as an interval of time since some particular event. For example, our implementation of `Instant` internally stores the number of nanoseconds since the moment when the year 1970 started in GMT. When `Instant` gets printed, it is displayed as the moment in the GMT time zone.

> why we still need to pass a TimeZone when asking the question "What is the absolute moment in time".

Let's say I look at the clock hanging on my wall and record the time when I started writing this: `LocalDateTime(2020, 12, 16, hours = 16, minutes = 20)`. Note that `LocalDateTime` doesn't store the time zone in which my clock operates! Then, if I want to tell you when I started writing this response, I can't just say "16:20", as your time zone is likely different; likewise, I can't just send my `LocalDateTime` instance to you. Instead, I convert it to an `Instant` and say "I'm writing this 1608124800 seconds after the year 1970 began in the GMT time zone"–an objective truth that doesn't depend at all on where I'm writing this.
```kotlin
        val ldt = LocalDateTime(2020, 12, 16, 16, 20)
        val instant = ldt.toInstant(TimeZone.of("Europe/Moscow"))
        println(instant) // 2020-12-16T13:20:00Z
        println(instant.epochSeconds) // 1608124800
```

This is somewhat simplified and doesn't mention leap seconds, but I was going for the general idea in this explanation, not the absolute correctness.
